src/token-getter.js

```diff
@@ -0,0 +1,68 @@
+'use strict';
+const AWS = require('aws-sdk');
+const clui = require('clui');
+
+class TokenGetter {
+  constructor(config, samlAssertion, account, role) {
+    this.spinner = new clui.Spinner('Getting token...');
+    this.sts = new AWS.STS({region: config.region});
+    this.samlAssertion = samlAssertion;
+    this.role = role;
+    this.account = account;
+    this.defaultAccount = config.defaultAccount;
+    this.accountNumber = this.account.accountNumber;
+  }
+
+  async getToken() {
+    try {
+      this.spinner.start();
+      const token = await this.getSTSToken();
+
+      if (this.isDefaultAccount()) {
+        this.spinner.stop();
+        return token;
+      }
+
+      // need to switch roles to the other account
+      const assumedToken = await this.getAssumeRoleToken(token);
+      this.spinner.stop();
+      return assumedToken;
+    } catch (e) {
+      console.log(e.stack); // eslint-disable-line no-console
+      throw new Error('error getting token: ' + e.message);
+    }
+  }
+
+  isDefaultAccount() {
+    return this.account.name === this.defaultAccount;
+  }
+
+  async getSTSToken() {
+    const request = this.sts.assumeRoleWithSAML({
+      PrincipalArn: this.role.principalArn,
+      RoleArn: this.role.roleArn,
+      SAMLAssertion: this.samlAssertion
+    });
+    return await request.promise();
+  }
+
+  async getAssumeRoleToken(originalToken) {
+    this.sts.config.credentials = new AWS.Credentials(
+      originalToken.Credentials.AccessKeyId,
+      originalToken.Credentials.SecretAccessKey,
+      originalToken.Credentials.SessionToken);
+    const roleArn = this.role.roleArn.replace(/::(\d+)/, `::${this.accountNumber}`);
+    const splitArn = originalToken.AssumedRoleUser.Arn.split('/');
+
+    return await this.sts.assumeRole({
+      RoleArn: roleArn,
+      RoleSessionName: splitArn[splitArn.length - 1]
+    }).promise();
+  }
+}
+
+function *getToken(config, samlAssertion, account, role) {
+  return new TokenGetter(config, samlAssertion, account, role).getToken();
```

Comment:
It seems not great that this function constructs a new TokenGetter each time just to call `getToken` on it with no args. It'd be nicer if the code presently calling it created a TokenGetter given args required to construct it and then called `getToken` with the args required to get a token for a particular account or role.

Reply:
Here the getToken function is used as a factory method, to keep the interface as similar as possible to how it was.
I could instantiate the TokenGetter in the index.js but will increase coupling to its implementation.

Reply:
Importing and instantiating TokenGetter in index now.